File: jar/src/main/java/jline/lib/perm/Permanent.java
/*
 * Copyright (c) 2012-2026, QORE Lab, Imperial College London
 * All rights reserved.
 */
package jline.lib.perm;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import jline.util.Pair;
import jline.util.matrix.Matrix;

/**
 * Implementation of the MATLAB perm.m permanent computation algorithm.
 */
public class Permanent extends PermSolver {

    public Permanent(Matrix matrix) {
        this(matrix, false);
    }

    public Permanent(Matrix matrix, boolean solve) {
        super(matrix);
        if (solve) {
            solve();
        }
    }

    @Override
    public void compute() {
        value = computeWithMultiplicities();
    }

    /**
     * Computes the permanent using the MATLAB perm.m algorithm with multiplicities.
     */
    private double computeWithMultiplicities() {
        Pair<Matrix, int[]> uniqueResult = findUniqueColumnsWithMultiplicities();
        Matrix uniqueMatrix = uniqueResult.getLeft();
        int[] multiplicities = uniqueResult.getRight();

        int R = multiplicities.length;
        int n = 0;
        for (int m : multiplicities) n += m;
        double result = 0.0;

        int[] f = new int[R];

        boolean done = false;
        while (!done) {
            int fSum = 0;
            for (int v : f) fSum += v;
            double term = Math.pow(-1.0, fSum);

            for (int j = 0; j < R; j++) {
                term *= binomialCoefficient(multiplicities[j], f[j]);
            }

            for (int i = 0; i < n; i++) {
                double sumTerm = 0.0;
                for (int k = 0; k < R; k++) {
                    sumTerm += f[k] * uniqueMatrix.get(i, k);
                }
                term *= sumTerm;
            }

            result += term;
            f = pprodNext(f, multiplicities);
            if (f.length == 0) {
                done = true;
            } else {
                boolean allMinus1 = true;
                for (int v : f) {
                    if (v != -1) { allMinus1 = false; break; }
                }
                if (allMinus1) done = true;
            }
        }

        return Math.pow(-1.0, n) * result;
    }

    /**
     * Log10 magnitude of the largest Ryser term for this orientation.
     *
     * The inclusion-exclusion expansion is largest when every column is selected,
     * giving prod_i (sum_j a_ij). Since per(A) equals per(A transposed), the two
     * orientations return the same value but not the same cancellation, so this
     * is the quantity to minimize when choosing between them. Returns positive
     * infinity when a row sum vanishes, so such an orientation is never chosen.
     */
    private static double ryserConditioning(Matrix m) {
        double total = 0.0;
        for (int i = 0; i < m.getNumRows(); i++) {
            double lineSum = 0.0;
            for (int j = 0; j < m.getNumCols(); j++) {
                lineSum += Math.abs(m.get(i, j));
            }
            if (lineSum <= 0.0) {
                return Double.POSITIVE_INFINITY;
            }
            total += Math.log10(lineSum);
        }
        return total;
    }

    /**
     * Groups repeated columns of the better-conditioned orientation of the matrix.
     *
     * Exploiting repeated rows means expanding the transpose, which leaves the
     * permanent unchanged but can raise the largest intermediate term by many
     * orders of magnitude. Vandermonde-like matrices, such as the A_x of
     * pfqn_lcfsqn_nc whose rows are geometric in the column index, lose every
     * significant digit that way. Orientation is chosen by conditioning first
     * and grouping applied second, even when that forgoes the grouping.
     */
    private Pair<Matrix, int[]> findUniqueColumnsWithMultiplicities() {
        Matrix source = matrix;
        Matrix transposed = matrix.transpose();
        if (ryserConditioning(transposed) < ryserConditioning(matrix)) {
            source = transposed;
        }

        Map<List<Double>, Integer> columnMap = new HashMap<List<Double>, Integer>();
        List<List<Double>> uniqueColumns = new ArrayList<List<Double>>();
        List<Integer> multiplicities = new ArrayList<Integer>();

        for (int j = 0; j < source.getNumCols(); j++) {
            List<Double> column = new ArrayList<Double>(source.getNumRows());
            for (int i = 0; i < source.getNumRows(); i++) {
                column.add(source.get(i, j));
            }
            Integer index = columnMap.get(column);
            if (index == null) {
                columnMap.put(column, uniqueColumns.size());
                uniqueColumns.add(column);
                multiplicities.add(1);
            } else {
                multiplicities.set(index, multiplicities.get(index) + 1);
            }
        }

        Matrix uniqueMatrix = new Matrix(source.getNumRows(), uniqueColumns.size());
        for (int j = 0; j < uniqueColumns.size(); j++) {
            List<Double> col = uniqueColumns.get(j);
            for (int i = 0; i < col.size(); i++) {
                uniqueMatrix.set(i, j, col.get(i));
            }
        }
        int[] arr = new int[multiplicities.size()];
        for (int k = 0; k < arr.length; k++) arr[k] = multiplicities.get(k);
        return new Pair<Matrix, int[]>(uniqueMatrix, arr);
    }

    /**
     * Computes binomial coefficient C(n, k).
     */
    private double binomialCoefficient(int n, int k) {
        if (k > n || k < 0) return 0.0;
        if (k == 0 || k == n) return 1.0;

        double result = 1.0;
        int upper = Math.min(k, n - k);
        for (int i = 1; i <= upper; i++) {
            result = result * (n - i + 1) / i;
        }
        return result;
    }

    /**
     * MATLAB pprod iterator - generates the next state in the sequence.
     */
    private int[] pprodNext(int[] current, int[] bounds) {
        int[] n = current.clone();
        int[] N = bounds;
        int R = N.length;

        boolean atMax = true;
        for (int i = 0; i < R; i++) {
            if (n[i] != N[i]) { atMax = false; break; }
        }
        if (atMax) {
            return new int[] { -1 };
        }

        int s = R - 1;
        while (s >= 0 && n[s] == N[s]) {
            n[s] = 0;
            s--;
        }

        if (s < 0) {
            return new int[] { -1 };
        }

        n[s]++;
        return n;
    }
}

File: jar/src/main/kotlin/jline/lib/perm/Permanent.kt
package jline.lib.perm

import jline.util.matrix.Matrix
import kotlin.math.pow

/**
 * Implementation of the MATLAB perm.m permanent computation algorithm.
 * 
 * This implementation mirrors the MATLAB perm.m function which computes the permanent
 * of a matrix by applying computational savings when some rows or columns are repeated.
 * It uses the inclusion-exclusion principle with multiplicities to efficiently handle
 * matrices with duplicate rows or columns.
 * 
 * The algorithm:
 * 1. Detects and counts repeated columns/rows
 * 2. Uses inclusion-exclusion with multinomial coefficients
 * 3. Applies pprod iterator for generating all combinations
 * 
 * @param matrix The matrix for which to compute the permanent
 * @param solve Whether to automatically run solve() after construction (default: false)
 */
class Permanent(
    matrix: Matrix,
    solve: Boolean = false
) : PermSolver(matrix) {
    
    init {
        if (solve) {
            solve()
        }
    }
    
    override fun compute() {
        value = computeWithMultiplicities()
    }
    
    /**
     * Computes the permanent using the MATLAB perm.m algorithm with multiplicities.
     * 
     * @return The permanent value
     */
    private fun computeWithMultiplicities(): Double {
        // Find unique columns and their multiplicities
        val (uniqueMatrix, multiplicities) = findUniqueColumnsWithMultiplicities()
        
        val R = multiplicities.size
        val n = multiplicities.sum()
        var value = 0.0
        
        // Initialize pprod iterator
        var f = IntArray(R) { 0 }
        
        do {
            var term = (-1.0).pow(f.sum())
            
            // Multinomial coefficients
            for (j in 0 until R) {
                term *= binomialCoefficient(multiplicities[j], f[j])
            }
            
            // Product term
            for (i in 0 until n) {
                var sumTerm = 0.0
                for (k in 0 until R) {
                    sumTerm += f[k] * uniqueMatrix.get(i, k)
                }
                term *= sumTerm
            }
            
            value += term
            f = pprodNext(f, multiplicities)
        } while (f.isNotEmpty() && !f.all { it == -1 })
        
        return (-1.0).pow(n) * value
    }
    
    /**
     * Finds unique columns in the matrix and their multiplicities.
     * If no repeated columns, checks for repeated rows instead.
     * 
     * @return Pair of (unique matrix, multiplicities array)
     */
    private fun findUniqueColumnsWithMultiplicities(): Pair<Matrix, IntArray> {
        // Find unique columns
        val columnMap = mutableMapOf<List<Double>, Int>()
        val uniqueColumns = mutableListOf<List<Double>>()
        val multiplicities = mutableListOf<Int>()
        
        for (j in 0 until matrix.getNumCols()) {
            val column = (0 until matrix.getNumRows()).map { i -> matrix.get(i, j) }
            val index = columnMap[column]
            if (index == null) {
                columnMap[column] = uniqueColumns.size
                uniqueColumns.add(column)
                multiplicities.add(1)
            } else {
                multiplicities[index]++
            }
        }
        
        // If no repeated columns (all multiplicities are 1), check for repeated rows
        if (multiplicities.all { it == 1 }) {
            val rowMap = mutableMapOf<List<Double>, Int>()
            val uniqueRows = mutableListOf<List<Double>>()
            val rowMultiplicities = mutableListOf<Int>()
            
            for (i in 0 until matrix.getNumRows()) {
                val row = (0 until matrix.getNumCols()).map { j -> matrix.get(i, j) }
                val index = rowMap[row]
                if (index == null) {
                    rowMap[row] = uniqueRows.size
                    uniqueRows.add(row)
                    rowMultiplicities.add(1)
                } else {
                    rowMultiplicities[index]++
                }
            }
            
            // If there are repeated rows, transpose and use row multiplicities
            if (rowMultiplicities.any { it > 1 }) {
                val transposedMatrix = Matrix(uniqueRows[0].size, uniqueRows.size)
                for (i in uniqueRows.indices) {
                    for (j in uniqueRows[i].indices) {
                        transposedMatrix.set(j, i, uniqueRows[i][j])
                    }
                }
                return Pair(transposedMatrix, rowMultiplicities.toIntArray())
            } else {
                // No repeated columns or rows, use original unique columns
                val uniqueMatrix = Matrix(matrix.getNumRows(), uniqueColumns.size)
                for (j in uniqueColumns.indices) {
                    for (i in uniqueColumns[j].indices) {
                        uniqueMatrix.set(i, j, uniqueColumns[j][i])
                    }
                }
                return Pair(uniqueMatrix, multiplicities.toIntArray())
            }
        } else {
            // There are repeated columns, use unique columns
            val uniqueMatrix = Matrix(matrix.getNumRows(), uniqueColumns.size)
            for (j in uniqueColumns.indices) {
                for (i in uniqueColumns[j].indices) {
                    uniqueMatrix.set(i, j, uniqueColumns[j][i])
                }
            }
            return Pair(uniqueMatrix, multiplicities.toIntArray())
        }
    }
    
    /**
     * Computes binomial coefficient C(n, k).
     * 
     * @param n The upper parameter
     * @param k The lower parameter
     * @return The binomial coefficient
     */
    private fun binomialCoefficient(n: Int, k: Int): Double {
        if (k > n || k < 0) return 0.0
        if (k == 0 || k == n) return 1.0
        
        var result = 1.0
        for (i in 1..minOf(k, n - k)) {
            result = result * (n - i + 1) / i
        }
        return result
    }
    
    /**
     * MATLAB pprod iterator - generates the next state in the sequence.
     * Returns a sequence of non-negative vectors less than a given vector.
     * 
     * @param current Current state vector
     * @param bounds Upper bounds vector
     * @return Next state vector, or array of -1s if sequence is complete
     */
    private fun pprodNext(current: IntArray, bounds: IntArray): IntArray {
        val n = current.copyOf()
        val N = bounds
        val R = N.size
        
        // Check if we've reached the maximum state
        if (n.zip(N).all { (a, b) -> a == b }) {
            return intArrayOf(-1)
        }
        
        var s = R - 1
        while (s >= 0 && n[s] == N[s]) {
            n[s] = 0
            s--
        }
        
        if (s < 0) {
            return intArrayOf(-1)
        }
        
        n[s]++
        return n
    }
}